File: 189_Rotate Array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # 0. 2-line short-cut
        k = k % len(nums)
        nums[:] = nums[-k:] + nums[:-k]

        # # 1. k % len(nums) to minimize rotate steps -> no change if 0
        # k = k % len(nums)
        # if k == 0: return

        # def mirror(left: int, right: int):
        #     while left < right:
        #         nums[left], nums[right] = nums[right], nums[left]
        #         left += 1
        #         right -= 1
        
        # # 2. mirror nums[:]
        # mirror(0, len(nums) - 1)
        
        # # 3. mirror nums[:k) and nums[k:]
        # mirror(0, k-1)
        # mirror(k, len(nums)-1)

File: 189_Rotate Array.java
// 时间复杂度：O(n)，其中 n 是 nums 的长度。
// 空间复杂度：O(1)。

class Solution {
    private void mirror(int[] nums, int left, int right) {
        while (left < right) {
            int tmp = nums[left];
            nums[left++] = nums[right];
            nums[right--] = tmp;
        }
    }

    public void rotate(int[] nums, int k) {
        // 1. min k
        k = k % nums.length;
        
        // 2. mirror nums[:]
        mirror(nums, 0, nums.length - 1);
        
        // 3. mirror nums[:k) and nums[k:]
        mirror(nums, 0, k - 1);
        mirror(nums, k, nums.length - 1);
    }
}